runtime/unix.js

```diff
@@ -64,3 +64,180 @@ function unix_isatty(fileDescriptor) {
     return false;
   }
 }
+
+//Provides: caml_raise_unix_exn_of_nodejs_error
+//Requires: caml_raise_with_args
+//Requires: caml_named_value, caml_string_of_jsbytes
+function caml_raise_unix_exn_of_nodejs_error(err) {
+  /* ===Unix.error===
+   *
+   * This array is in order of the variant in OCaml
+   */
+  var errors = [
+    'E2BIG', 'EACCES', 'EAGAIN', 'EBADF', 'EBUSY', 'ECHILD', 'EDEADLK', 'EDOM',
+    'EEXIST', 'EFAULT', 'EFBIG', 'EINTR', 'EINVAL', 'EIO', 'EISDIR', 'EMFILE',
+    'EMLINK', 'ENAMETOOLONG', 'ENFILE', 'ENODEV', 'ENOENT', 'ENOEXEC', 'ENOLCK',
+    'ENOMEM', 'ENOSPC', 'ENOSYS', 'ENOTDIR', 'ENOTEMPTY', 'ENOTTY', 'ENXIO',
+    'EPERM', 'EPIPE', 'ERANGE', 'EROFS', 'ESPIPE', 'ESRCH', 'EXDEV', 'EWOULDBLOCK',
+    'EINPROGRESS', 'EALREADY', 'ENOTSOCK', 'EDESTADDRREQ', 'EMSGSIZE',
+    'EPROTOTYPE', 'ENOPROTOOPT', 'EPROTONOSUPPORT', 'ESOCKTNOSUPPORT',
+    'EOPNOTSUPP', 'EPFNOSUPPORT', 'EAFNOSUPPORT', 'EADDRINUSE', 'EADDRNOTAVAIL',
+    'ENETDOWN', 'ENETUNREACH', 'ENETRESET', 'ECONNABORTED', 'ECONNRESET', 'ENOBUFS',
+    'EISCONN', 'ENOTCONN', 'ESHUTDOWN', 'ETOOMANYREFS', 'ETIMEDOUT', 'ECONNREFUSED',
+    'EHOSTDOWN', 'EHOSTUNREACH', 'ELOOP', 'EOVERFLOW'
+  ];
+  var variant = errors.indexOf(err.code);
+  if (variant < 0) {
+    // TODO: Make sure this works `EUNKNOWNERR(int)`
+    variant = BLOCK(0, err.errno);
+  }
+  var args = [
+    variant,
+    caml_string_of_jsbytes(err.syscall || ''),
+    caml_string_of_jsbytes(err.path || '')
+  ];
+  caml_raise_with_args(caml_named_value('Unix.Unix_error'), args.length, args);
+}
+
+//Provides: caml_stats_from_js
+function caml_stats_from_js(js_stats) {
+  /* ===Unix.file_kind===
+   * type file_kind =
+   *     S_REG                       (** Regular file *)
+   *   | S_DIR                       (** Directory *)
+   *   | S_CHR                       (** Character device *)
+   *   | S_BLK                       (** Block device *)
+   *   | S_LNK                       (** Symbolic link *)
+   *   | S_FIFO                      (** Named pipe *)
+   *   | S_SOCK                      (** Socket *)
+   */
+  var file_kind;
+  if (js_stats.isFile()) {
+    file_kind = 0;
+  }
+  if (js_stats.isDirectory()) {
```

Comment:
Are we sure all those ifs are independent ? I would probably use if-else-if ..

Reply:
Done in my refactor. I believe they are all independent, but we can be careful.